File: devswitch/service.py
# -*- coding: utf-8 -*-
from __future__ import print_function, unicode_literals

import os
from pathlib import Path
from typing import List, Optional, Tuple

from . import apply, detect, paths
from .models import Runtime, State, TOOLS, TOOL_LABELS
from .store import load_state, save_state


def merge_scan(state=None):
    # type: (Optional[State]) -> State
    state = state or load_state()
    scanned = detect.scan_runtimes()
    seen_homes = set()
    merged = []
    for runtime in scanned:
        key = (runtime.tool, runtime.home.rstrip("/"))
        if key in seen_homes:
            continue
        seen_homes.add(key)
        previous = state.by_home(runtime.tool, runtime.home)
        if previous and previous.source == "imported":
            runtime.source = "imported"
            runtime.label = previous.label or runtime.label
        merged.append(runtime)
    for runtime in state.runtimes:
        key = (runtime.tool, runtime.home.rstrip("/"))
        if key in seen_homes:
            continue
        if runtime.source == "imported" and Path(runtime.home).exists():
            refreshed = (
                detect.inspect_node(Path(runtime.home))
                if runtime.tool == "node"
                else detect.inspect_java(Path(runtime.home))
            )
            if refreshed is not None:
                refreshed.source = "imported"
                merged.append(refreshed)
                seen_homes.add(key)
    state.runtimes = merged
    for tool in TOOLS:
        current = state.current_runtime(tool)
        if current is None:
            guessed = detect.infer_active(state.runtimes, tool)
            state.current[tool] = guessed.home if guessed is not None else ""
    save_state(state)
    return state


def use(tool, query, state=None):
    # type: (str, str, Optional[State]) -> Runtime
    state = state or load_state()
    runtime = state.find(tool, query)
    if runtime is None:
        state = merge_scan(state)
        runtime = state.find(tool, query)
    if runtime is None:
        if not state.for_tool(tool):
            raise LookupError(
                "本机没有发现 {}。安装后执行 devswitch scan，或：devswitch import {} <目录>".format(
                    tool, tool
                )
            )
        raise LookupError("没有找到 {} 版本：{}".format(tool, query))
    apply.apply_switch(state, tool, runtime)
    apply.apply_hooks(state)
    return runtime


def import_path(tool, path):
    # type: (str, str) -> Runtime
    home = Path(os.path.expanduser(path)).resolve()
    inspector = {
        "node": detect.inspect_node,
        "java": detect.inspect_java,
        "maven": detect.inspect_maven,
        "gradle": detect.inspect_gradle,
    }.get(tool)
    if inspector is None:
        raise ValueError("不支持的工具：{}".format(tool))
    runtime = inspector(home)
    if runtime is None:
        raise LookupError("目录里没有可用的 {}：{}".format(tool, home))
    runtime.source = "imported"
    state = load_state()
    state.upsert(runtime)
    save_state(state)
    if state.current_runtime(tool) is None:
        apply.apply_switch(state, tool, runtime)
        apply.apply_hooks(state)
    return runtime


def ensure_applied(state=None):
    # type: (Optional[State]) -> State
    state = state or load_state()
    apply.apply_all(state)
    return state


def export_versions(path):
    # type: (str) -> str
    """Write the selected versions to a shareable .devswitch file."""
    import json

    state = load_state()
    tools = {}
    for tool in TOOLS:
        runtime = state.current_runtime(tool)
        if runtime is not None:
            tools[tool] = runtime.version
    payload = {"devswitch": 1, "tools": tools}
    Path(path).write_text(
        json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
    )
    return str(Path(path))


def apply_versions(path, install_missing=False, mirror=None):
    # type: (str, bool, Optional[str]) -> List[Tuple[str, str, str]]
    """Apply a .devswitch file: switch each tool to the recorded version.
    Returns [(tool, version, result)] with result in
    switched / installed / missing-version / unknown-tool / failed: …"""
    import json

    try:
        payload = json.loads(Path(path).read_text(encoding="utf-8"))
    except (OSError, ValueError) as exc:
        raise ValueError("无法读取 {}: {}".format(path, exc))
    results = []
    for tool, version in ((payload or {}).get("tools") or {}).items():
        if tool not in TOOLS:
            results.append((tool, version, "unknown-tool"))
            continue
        try:
            use(tool, version)
            results.append((tool, version, "switched"))
        except LookupError:
            if not install_missing:
                results.append((tool, version, "missing-version"))
                continue
            try:
                from . import downloader

                downloader.install_runtime(tool, version.split(".")[0], mirror)
                results.append((tool, version, "installed"))
            except (LookupError, ValueError, OSError) as exc:
                results.append((tool, version, "failed: {}".format(exc)))
    return results


WINDOWS_NODE_TARGETS = {
    "node": "node.exe",
    "npm": "npm.cmd",
    "npx": "npx.cmd",
    "corepack": "corepack.cmd",
}
WINDOWS_BUILDTOOL_TARGETS = {
    "mvn": "mvn.cmd",
    "mvnDebug": "mvnDebug.cmd",
    "gradle": "gradle.bat",
}


def which_binary(name):
    # type: (str) -> Optional[str]
    state = load_state()
    # <tool>-home / <TOOL>_HOME / 大写工具名 → home 目录；裸命令名走下面的二进制分支
    for tool in TOOLS:
        if name in (tool + "-home", tool.upper() + "_HOME", tool.upper()):
            runtime = state.current_runtime(tool)
            return runtime.home if runtime else None
    if name in ("node", "npm", "npx", "corepack"):
        runtime = state.current_runtime("node")
        if runtime is None:
            return None
        if paths.IS_WINDOWS:
            candidate = Path(runtime.home) / WINDOWS_NODE_TARGETS[name]
        else:
            candidate = Path(runtime.home) / "bin" / name
        return str(candidate) if candidate.exists() else None
    if name in ("mvn", "mvnDebug", "gradle"):
        tool = "maven" if name.startswith("mvn") else "gradle"
        runtime = state.current_runtime(tool)
        if runtime is None:
            return None
        if paths.IS_WINDOWS:
            candidate = Path(runtime.home) / "bin" / WINDOWS_BUILDTOOL_TARGETS[name]
        else:
            candidate = Path(runtime.home) / "bin" / name
        return str(candidate) if candidate.exists() else None
    runtime = state.current_runtime("java")
    if runtime is None:
        return None
    exe = name + ".exe" if paths.IS_WINDOWS else name
    for rel in ("bin/" + exe, "jre/bin/" + exe):
        candidate = Path(runtime.home) / rel
        if candidate.exists():
            return str(candidate)
    return None


def _norm(text):
    # type: (str) -> str
    return os.path.normcase(text) if paths.IS_WINDOWS else text


def doctor_issues():
    # type: () -> List[Tuple[str, str, str]]
    """Return list of (level, code, message)."""
    issues = []
    state = load_state()
    local_bin = str(paths.local_bin())
    path_env = os.environ.get("PATH") or ""
    parts = [item for item in path_env.split(os.pathsep) if item]
    norm_parts = [_norm(item) for item in parts]

    if _norm(local_bin) not in norm_parts:
        if paths.IS_WINDOWS:
            issues.append(
                (
                    "warn",
                    "path-missing",
                    "{} 不在本次 PATH 里，终端可能仍走系统版本。".format(local_bin),
                )
            )
        else:
            issues.append(
                (
                    "warn",
                    "path-missing",
                    "~/.local/bin 不在 PATH 里，图形程序和部分终端可能仍走系统版本。",
                )
            )
    elif norm_parts and norm_parts[0] != _norm(local_bin):
        keywords = ("node", "jvm", "jdk", "java") if paths.IS_WINDOWS else ("node", "jvm")
        ahead = [
            item
            for index, item in enumerate(parts[: norm_parts.index(_norm(local_bin))])
            if any(word in _norm(item) for word in keywords)
        ]
        if ahead:
            if paths.IS_WINDOWS:
                issues.append(
                    (
                        "warn",
                        "path-order",
                        "PATH 里有更靠前的 Node/Java 目录（常见于机器级安装），可能抢在 DevSwitch shim 前面：{}".format(
                            ", ".join(ahead)
                        ),
                    )
                )
            else:
                issues.append(
                    (
                        "warn",
                        "path-order",
                        "PATH 里有更靠前的 Node/Java 目录，可能抢在 DevSwitch shim 前面：{}".format(
                            ", ".join(ahead)
                        ),
                    )
                )

    if paths.IS_WINDOWS:
        from . import winenv

        user_entries = [_norm(item) for item in winenv.user_path_entries()]
        if _norm(local_bin) not in user_entries:
            issues.append(
                (
                    "warn",
                    "user-path-missing",
                    "{} 不在用户 PATH（注册表）里，新开的终端会找不到 shim。可运行 devswitch hook install。".format(
                        local_bin
                    ),
                )
            )
        for tool, var in apply.TOOL_HOME_VARS.items():
            runtime = state.current_runtime(tool)
            if runtime is None:
                continue
            reg_home = winenv.get_user_value(var)
            reg_home = reg_home[0] if reg_home else None
            if not reg_home or _norm(os.path.normpath(reg_home)) != _norm(os.path.normpath(runtime.home)):
                issues.append(
                    (
                        "warn",
                        tool + "-home-registry",
                        "用户环境变量 {}（注册表）与当前选中版本不一致：{}。".format(
                            var, reg_home or "未设置"
                        ),
                    )
                )

    env_probe = paths.env_file() if paths.IS_WINDOWS else paths.current_env_file()
    if not env_probe.exists():
        issues.append(("error", "no-env", "还没有生成切换配置，先运行 devswitch scan。"))

    probe_shims = {"node": ("node", "npm"), "java": ("java", "javac"),
                   "maven": ("mvn",), "gradle": ("gradle",)}
    shim_names = []
    for tool, names in probe_shims.items():
        if state.current_runtime(tool) is not None:
            shim_names.extend(names)
    for name in shim_names:
        shim = paths.local_bin() / paths.shim_filename(name)
        if not shim.exists():
            issues.append(("warn", "no-shim", "缺少 shim：{}".format(shim)))
        elif shim.is_symlink() or "DevSwitch shim" not in shim.read_text(encoding="utf-8", errors="ignore"):
            issues.append(("warn", "foreign-shim", "{} 还不是 DevSwitch 接管的入口。".format(shim)))

    if paths.IS_WINDOWS:
        hook_targets = paths.hook_targets()
        missing = [str(item) for item in hook_targets if not apply.hook_installed(item)]
        if len(missing) == len(hook_targets):
            issues.append(
                (
                    "warn",
                    "no-hook",
                    "还没有写入 PowerShell / Git Bash 钩子，新开终端可能没有 JAVA_HOME，也可能抢不过机器级安装。",
                )
            )
    elif not apply.hook_installed(paths.bashrc()):
        issues.append(("warn", "no-hook", "还没有写入 ~/.bashrc 钩子，新开终端可能没有 JAVA_HOME。"))

    for file_path, line in apply.conflicting_bashrc_lines():
        issues.append(
            (
                "warn",
                "bashrc-path",
                "{} 里仍有手写 Node PATH，容易和切换打架：{}".format(file_path, line),
            )
        )

    labels = TOOL_LABELS
    for tool in TOOLS:
        if not state.for_tool(tool):
            issues.append(
                (
                    "info",
                    "no-runtime",
                    "本机没有发现 {}。安装后执行扫描，或用导入指定目录。".format(labels[tool]),
                )
            )
        elif state.current_runtime(tool) is None:
            issues.append(("info", "no-current", "尚未选择当前 {} 版本。".format(labels[tool])))

    return issues


def doctor_fix():
    # type: () -> List[str]
    actions = []
    state = merge_scan()
    apply.apply_all(state)
    if paths.IS_WINDOWS:
        actions.append("已写入 shim、env.sh、PowerShell / Git Bash 钩子和用户环境变量")
    else:
        actions.append("已写入 shim、env.sh 和 shell 钩子")
    changed = apply.comment_conflicting_path_lines(paths.bashrc())
    changed += apply.comment_conflicting_path_lines(paths.profile())
    if changed:
        actions.append("已注释冲突 PATH：" + "；".join(changed))
    return actions

File: internal/service/service.go
// Package service 承载业务编排，平移自 devswitch/service.py。
package service

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/xyztony999/devswitch/internal/apply"
	"github.com/xyztony999/devswitch/internal/detect"
	"github.com/xyztony999/devswitch/internal/downloader"
	"github.com/xyztony999/devswitch/internal/models"
	"github.com/xyztony999/devswitch/internal/paths"
	"github.com/xyztony999/devswitch/internal/store"
)

func MergeScan() models.State {
	state := store.Load()
	scanned := detect.ScanRuntimes()
	seen := map[string]bool{}
	merged := []models.Runtime{}
	for i := range scanned {
		key := string(scanned[i].Tool) + ":" + strings.TrimRight(scanned[i].Home, "/")
		if seen[key] {
			continue
		}
		seen[key] = true
		if previous := state.ByHome(scanned[i].Tool, scanned[i].Home); previous != nil && previous.Source == "imported" {
			scanned[i].Source = "imported"
			if previous.Label != "" {
				scanned[i].Label = previous.Label
			}
		}
		merged = append(merged, scanned[i])
	}
	for i := range state.Runtimes {
		r := state.Runtimes[i]
		key := string(r.Tool) + ":" + strings.TrimRight(r.Home, "/")
		if seen[key] {
			continue
		}
		if r.Source == "imported" && fileExists(r.Home) {
			inspector := detect.Inspectors()[r.Tool]
			if inspector != nil {
				if refreshed := inspector(r.Home); refreshed != nil {
					refreshed.Source = "imported"
					merged = append(merged, *refreshed)
					seen[key] = true
				}
			}
		}
	}
	state.Runtimes = merged
	for _, tool := range models.Tools {
		if state.CurrentRuntime(tool) == nil {
			if guessed := detect.InferActive(state.Runtimes, tool); guessed != nil {
				state.Current[tool] = guessed.Home
			} else {
				state.Current[tool] = ""
			}
		}
	}
	_ = store.Save(state)
	return state
}

func fileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

// Use 切换版本并让全部生效动作落地。
func Use(tool models.Tool, query string) (*models.Runtime, error) {
	state := store.Load()
	runtime := state.Find(tool, query)
	if runtime == nil {
		state = MergeScan()
		runtime = state.Find(tool, query)
	}
	if runtime == nil {
		if len(state.ForTool(tool)) == 0 {
			return nil, fmt.Errorf("本机没有发现 %s。安装后执行 devswitch scan，或：devswitch install %s <版本>", tool, tool)
		}
		return nil, fmt.Errorf("没有找到 %s 版本：%s", tool, query)
	}
	state.Current[tool] = runtime.Home
	_ = store.Save(state)
	applyAll(&state)
	return runtime, nil
}

// applyAll 生成一切生效物（shim/启动器/env/钩子/用户环境变量）。
func applyAll(state *models.State) {
	_ = apply.WriteCurrentEnv(state)
	_ = apply.WriteEnvSh(state)
	_, _ = apply.WriteShims(state)
	_ = apply.WriteLaunchers()
	applyHooks(state)
	ensureProcessPath()
}

// ensureProcessPath 把 shim 目录前置到本进程 PATH：
// 注册表/钩子只对新进程生效，正在运行的进程（尤其 GUI 里点「一键修复」后）
// 不会自动获得，doctor 的会话 PATH 检查会一直报旧状态。这里同步进程内环境。
func ensureProcessPath() {
	localBin := paths.LocalBin()
	parts := filepath.SplitList(os.Getenv("PATH"))
	for i, p := range parts {
		if normFold(p) == normFold(localBin) {
			if i == 0 {
				return
			}
			rest := append(append([]string{}, parts[:i]...), parts[i+1:]...)
			next := append([]string{localBin}, rest...)
			os.Setenv("PATH", strings.Join(next, string(os.PathListSeparator)))
			return
		}
	}
	os.Setenv("PATH", localBin+string(os.PathListSeparator)+os.Getenv("PATH"))
}

func applyHooks(state *models.State) {
	for _, target := range paths.HookTargets() {
		_, _ = apply.InstallHook(target)
	}
	writeUserEnv(state)
}

// EnsureApplied 对当前状态重新生成全部生效物。
func EnsureApplied() models.State {
	state := store.Load()
	applyAll(&state)
	return state
}

func writeUserEnv(state *models.State) {
	if !paths.IsWindows {
		return
	}
	writeUserEnvPlatform(state)
}

// ImportPath 导入一个本地安装目录。
func ImportPath(tool models.Tool, path string) (*models.Runtime, error) {
	home, err := filepath.Abs(os.ExpandEnv(path))
	if err != nil {
		home = path
	}
	inspector := detect.Inspectors()[tool]
	if inspector == nil {
		return nil, fmt.Errorf("不支持的工具：%s", tool)
	}
	runtime := inspector(home)
	if runtime == nil {
		return nil, fmt.Errorf("目录里没有可用的 %s：%s", tool, home)
	}
	runtime.Source = "imported"
	state := store.Load()
	state.Upsert(*runtime)
	_ = store.Save(state)
	if state.CurrentRuntime(tool) == nil {
		state.Current[tool] = runtime.Home
		_ = store.Save(state)
		applyAll(&state)
	}
	return runtime, nil
}

var winNodeTargets = map[string]string{
	"node": "node.exe", "npm": "npm.cmd", "npx": "npx.cmd", "corepack": "corepack.cmd",
}
var winBuildtoolTargets = map[string]string{
	"mvn": "mvn.cmd", "mvnDebug": "mvnDebug.cmd", "gradle": "gradle.bat",
}

// WhichBinary 返回当前选中工具的真实二进制 / home。
func WhichBinary(name string) (string, bool) {
	state := store.Load()
	// <tool>-home / <TOOL>_HOME / 大写工具名 → home
	for _, tool := range models.Tools {
		upper := strings.ToUpper(string(tool))
		if name == string(tool)+"-home" || name == upper+"_HOME" || name == upper {
			if r := state.CurrentRuntime(tool); r != nil {
				return r.Home, true
			}
			return "", false
		}
	}
	exists := func(p string) (string, bool) {
		if info, err := os.Stat(p); err == nil && !info.IsDir() {
			return p, true
		}
		return "", false
	}
	switch name {
	case "node", "npm", "npx", "corepack":
		r := state.CurrentRuntime(models.Node)
		if r == nil {
			return "", false
		}
		if paths.IsWindows {
			return exists(filepath.Join(r.Home, winNodeTargets[name]))
		}
		return exists(filepath.Join(r.Home, "bin", name))
	case "mvn", "mvnDebug", "gradle":
		tool := models.Maven
		if name == "gradle" {
			tool = models.Gradle
		}
		r := state.CurrentRuntime(tool)
		if r == nil {
			return "", false
		}
		if paths.IsWindows {
			return exists(filepath.Join(r.Home, "bin", winBuildtoolTargets[name]))
		}
		return exists(filepath.Join(r.Home, "bin", name))
	}
	r := state.CurrentRuntime(models.Java)
	if r == nil {
		return "", false
	}
	exe := name
	if paths.IsWindows {
		exe = name + ".exe"
	}
	for _, rel := range []string{filepath.Join("bin", exe), filepath.Join("jre", "bin", exe)} {
		if p, ok := exists(filepath.Join(r.Home, rel)); ok {
			return p, true
		}
	}
	return "", false
}

// ---------------------------------------------------------------------------
// export / apply
// ---------------------------------------------------------------------------

type devswitchFile struct {
	Devswitch int               `json:"devswitch"`
	Tools     map[string]string `json:"tools"`
}

func ExportVersions(path string) (string, error) {
	state := store.Load()
	tools := map[string]string{}
	for _, tool := range models.Tools {
		if r := state.CurrentRuntime(tool); r != nil {
			tools[string(tool)] = r.Version
		}
	}
	data, err := json.MarshalIndent(devswitchFile{Devswitch: 1, Tools: tools}, "", "  ")
	if err != nil {
		return "", err
	}
	if err := os.WriteFile(path, append(data, '\n'), 0o644); err != nil {
		return "", err
	}
	return path, nil
}

// InstallRuntime 下载安装并自动切换，返回 (home, 完整版本)。
func InstallRuntime(tool models.Tool, version, mirror string) (string, string, error) {
	inspector := detect.Inspectors()[tool]
	if inspector == nil {
		return "", "", fmt.Errorf("暂不支持下载该工具：%s", tool)
	}
	home, err := downloader.InstallTool(tool, version, mirror)
	if err != nil {
		return "", "", err
	}
	runtime := inspector(home)
	if runtime == nil {
		return "", "", fmt.Errorf("下载完成但未能识别安装目录：%s", home)
	}
	_ = MergeScan()
	if _, err := Use(tool, runtime.Version); err != nil {
		return "", "", err
	}
	return runtime.Home, runtime.Version, nil
}

type ApplyResult struct {
	Tool    string
	Version string
	Outcome string
}

func ApplyVersions(path string, installMissing bool, mirror string) ([]ApplyResult, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("无法读取 %s: %v", path, err)
	}
	var payload devswitchFile
	if err := json.Unmarshal(data, &payload); err != nil {
		return nil, fmt.Errorf("无法读取 %s: %v", path, err)
	}
	results := []ApplyResult{}
	// 固定顺序遍历（models.Tools），避免 map 迭代随机
	for _, tool := range models.Tools {
		version, ok := payload.Tools[string(tool)]
		if !ok {
			continue
		}
		if _, err := Use(tool, version); err == nil {
			results = append(results, ApplyResult{string(tool), version, "switched"})
			continue
		}
		if !installMissing {
			results = append(results, ApplyResult{string(tool), version, "missing-version"})
			continue
		}
		major := strings.Split(version, ".")[0]
		if _, _, err := InstallRuntime(tool, major, mirror); err != nil {
			results = append(results, ApplyResult{string(tool), version, "failed: " + err.Error()})
		} else {
			if _, err := Use(tool, version); err != nil {
				results = append(results, ApplyResult{string(tool), version, "failed: " + err.Error()})
			} else {
				results = append(results, ApplyResult{string(tool), version, "installed"})
			}
		}
	}
	return results, nil
}
